### PR TITLE
Cubic Spline Struct

### PR DESCRIPTION
**Pull request by schrieveslaach**
Add a struct that holds the vector of polynomials of a natural cubic spline, providing additionally an eval function that uses binary search to find the correct polynomial for a given x.

Additionally, the spline struct can be extended by new x and y values.

### CODE REVIEW

File: src/numerical/spline.rs [hunk 1]
```diff
@@ -36,71 +39,188 @@ use util::non_macro::*;
...
+    /// let x = c!(0.9, 1.3, 1.9, 2.1);
+    /// let y = c!(1.3, 1.5, 1.85, 2.1);
+    ///
+    /// let s = CubicSpline::from_x_and_y(x, y);
+    ///
+    /// for i in 0 .. 4 {
+    ///     println!("{}", s.eval(i as f64 / 2.0));
+    /// }
+    /// ```
+    pub fn from_x_and_y(node_x: Vector, node_y: Vector) -> Self {
```
Comment (schrieveslaach) [inline src/numerical/spline.rs]: @Axect, I'm not so happy about the function name. What is your opinion here?

File: src/numerical/spline.rs [hunk 2]
```diff
@@ -36,71 +39,188 @@ use util::non_macro::*;
...
+            }
+        }) {
+            Ok(index) => index,
+            Err(index) => max(0, min(index, self.polynomials.len() - 1)),
+        };
+
+        self.polynomials[index].1.eval(x)
+    }
+
+    pub fn extend_with_x_and_y(&mut self, node_x: Vector, node_y: Vector) {
```
Comment (schrieveslaach) [inline src/numerical/spline.rs]: Here I'm also not so happy.

File: src/numerical/spline.rs (hunk 1 above)
Comment (Axect) [reply src/numerical/spline.rs]: @schrieveslaach Hmm.. I'm also uncomfortable too. Wouldn't `from_nodes`be better?

File: src/numerical/spline.rs (hunk 2 above)
Comment (Axect) [reply src/numerical/spline.rs]: Also, it could be `extend_with_nodes`. What is your opinion about it?

File: src/numerical/spline.rs (hunk 1 above)
Comment (schrieveslaach) [reply src/numerical/spline.rs]: That sounds good. I'll fix that later.

**Comment by schrieveslaach** [comment]
@Axect, I've updated this PR with the new function and method names. I also added `From` and `Into` implementations.

**Comment by Axect** [comment]
Great! I'll merge it now.

**Comment by schrieveslaach** [comment]
Nice. :sunglasses:

**Pull request merged**